sequencer/src/block/full_payload/ns_proof.rs

```diff
@@ -0,0 +1,163 @@
+    /// Returns the payload bytes for namespace `ns_id`, along with a proof of
+    /// correctness for those bytes. Returns `None` if `ns_id` is not in the
+    /// namespace table, or on error.
+    ///
+    /// The namespace payload [`NsPayloadOwned`] is included as a hidden field
+    /// in the returned [`NsProof`]. A conventional API would instead return
+    /// `(NsPayload, NsProof)` and [`NsProof`] would not contain the namespace
+    /// payload.
+    /// ([`TxProof::new`](crate::block::namespace_payload::TxProof::new)
+    /// conforms to this convention.) In the future we should change this API to
+    /// conform to convention. But that would require a change to our RPC
+    /// endpoint API at [`endpoints`](crate::api::endpoints), which is a hassle.
+    pub fn new(payload: &Payload, index: &NsIndex, common: &VidCommon) -> Option<NsProof> {
+        let payload_byte_len = payload.byte_len();
+        payload_byte_len.is_consistent(common).ok()?;
+        if !payload.ns_table().in_bounds(index) {
+            return None; // error: index out of bounds
+        }
+        let ns_payload_range = payload.ns_table().ns_range(index, &payload_byte_len);
+
+        // TODO vid_scheme() arg should be u32 to match get_num_storage_nodes
+        let vid = vid_scheme(
+            VidSchemeType::get_num_storage_nodes(common)
+                .try_into()
+                .ok()?, // error: failure to convert u32 to usize
+        );
+
+        let ns_proof = if ns_payload_range.as_block_range().is_empty() {
+            None
+        } else {
+            Some(
+                vid.payload_proof(payload.encode(), ns_payload_range.as_block_range())
+                    .ok()?, // error: internal to payload_proof()
+            )
+        };
+
+        Some(NsProof {
+            ns_index: index.clone(),
+            ns_payload: payload.read_ns_payload(&ns_payload_range).to_owned(),
+            ns_proof,
+        })
+    }
+
+    /// Verify a [`NsProof`] against a payload commitment. Returns `None` on
+    /// error or if verification fails.
+    ///
+    /// There is no [`NsPayload`](crate::block::namespace_payload::NsPayload)
+    /// arg because this data is already included in the [`NsProof`]. See
+    /// [`NsProof::new`] for discussion.
+    ///
+    /// If verification is successful then return `(Vec<Transaction>,
+    /// NamespaceId)` obtained by post-processing the underlying
+    /// [`NsPayload`](crate::block::namespace_payload::NsPayload). Why? This
+    /// method might be run by a client in a WASM environment who might be
+    /// running non-Rust code, in which case the client is unable to perform
+    /// this post-processing himself.
+    pub fn verify(
+        &self,
+        ns_table: &NsTable,
+        commit: &VidCommitment,
+        common: &VidCommon,
+    ) -> Option<(Vec<Transaction>, NamespaceId)> {
+        VidSchemeType::is_consistent(commit, common).ok()?;
+        if !ns_table.in_bounds(&self.ns_index) {
```

Comment:
Nit: it seems like we often check `in_bounds` and then read from the index with `ns_range` or whatever. Maybe it would be cleaner to have `ns_range` do the bounds checking and return an `Option`? And if there are places where we already know the index is in range and want to skip the extra check for (dubious) efficiency, we could have `ns_range_unchecked`

Reply:
Fixed in https://github.com/EspressoSystems/espresso-sequencer/pull/1499/commits/176f0015e466ee3baf2a7ef473b617fbe003784b